File: MarsRoverFSharp/src/Rover.fs
namespace MarsRover

module Rover =
    open System

    let initialPosition = (0,0,"N")
    [<Literal>]
    let pattern = "{0}:{1}:{2}"
    [<Literal>]
    let Forward = 'F'
    [<Literal>]
    let Left = 'L'
    [<Literal>]
    let Right = 'R'
    [<Literal>]
    let North = "N";
    [<Literal>]
    let West = "W";
    [<Literal>]
    let South = "S";
    [<Literal>]
    let East = "E";

    let private Increment value =
        if value < 9 then
            value + 1
        else
            0

    let private Decrement value =
        if value = 0 then
            9
        else
            value - 1

    let private MoveForward position =
        let (x,y,orientation) = position
        match orientation with
        | North ->(x,Increment y,orientation)
        | East -> (Increment x,y,orientation)
        | South -> (x,Decrement y,orientation)
        |_ -> (Decrement x,y,orientation)
        
    let private FormatReturn position =
        let (x,y,orientation) = position
        String.Format (pattern, x, y, orientation) 
    
    let private GoLeft position =
        let (x,y,orientation) = position
        match orientation with
        | North -> (x,y,West)
        | West -> (x,y,South)
        | South -> (x,y,East)
        |_ -> (x,y,North)

    let private GoRight position =
        let (x,y,orientation) = position
        match orientation with
        | South -> (x,y,West)
        | East -> (x,y,South)
        | North -> (x,y,East)
        |_ -> (x,y,North)

    let private Action (direction, position) =
        match direction with
        | Left -> GoLeft position
        | Right -> GoRight position
        |_ -> MoveForward position

    let rec Execute (position, directions:string) =
        if directions.Length = 0 then
            position
        else
            let direction = directions.Chars(0)
            Execute(Action(direction , position), directions.Substring(1))

    let Move directions =
        Execute(initialPosition, directions) 
        |> FormatReturn        

File: MarsRoverCSharp/src/Rover.cs
using System.Linq;

namespace MarsRoverCSharp
{
    public class Rover
    {
        private const char Left = 'L';
        private const char Right = 'R';
        private const char Forward = 'F';
        private const string North = "N";
        private const string West = "W";
        private const string South = "S";
        private const string East = "E";
        private const int minimumValue = 0;
        private const int maximumValue = 9;

        private int abscisse = 0;
        private int ordonnnee = 0;
        private string orientation = North;

        public string Move(string directions)
        {
            directions
                .ToList()
                .ForEach(ExecuteDirection);
                                    
            return $"{abscisse}:{ordonnnee}:{orientation}";
        }

        private void ExecuteDirection(char direction)
        {
            switch (direction)
            {
                case Right:
                    GoRight();
                    break;
                case Left:
                    GoLeft();
                    break;
                case Forward:
                    GoForward();
                    break;
            }
        }

        private void GoForward()
        {
            switch (orientation)
            {
                case West:
                    abscisse = DontExitFromBoard(--abscisse);
                    break;
                case South:
                    ordonnnee = DontExitFromBoard(--ordonnnee);
                    break;
                case North:
                    ordonnnee = DontExitFromBoard(++ordonnnee);
                    break;
                case East:
                    abscisse = DontExitFromBoard(++abscisse);
                    break;
            }
        }

        private int DontExitFromBoard(int position)
        {
            if (position > maximumValue)
                return minimumValue;
            if (position < minimumValue)
                return maximumValue;
            return position;
        }

        private void GoLeft()
        {
            switch (orientation)
            {
                case West:
                    orientation = South;
                    break;
                case South:
                    orientation = East;
                    break;
                case North:
                    orientation = West;
                    break;
                case East:
                    orientation = North;
                    break;
            }
        }

        private void GoRight()
        {
            switch (orientation)
            {
                case East:
                    orientation = South;
                    break;
                case South:
                    orientation = West;
                    break;
                case North:
                    orientation = East;
                    break;
                case West:
                    orientation = North;
                    break;
            }
        }
    }
}
